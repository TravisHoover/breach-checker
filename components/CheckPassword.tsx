import React, { useState, useEffect, ReactElement } from 'react'
import { StyleSheet, TouchableOpacity } from 'react-native'
import sha1 from 'sha1'
import Colors from '../constants/Colors'
import { Text, TextInput, View } from './Themed'
import { AdMobBanner} from "expo-ads-admob";
import Constants from 'expo-constants';

export default function CheckPassword (): ReactElement {
  const [password, setPassword] = useState('')
  const [results, setResults] = useState('')
  const testID = 'ca-app-pub-3940256099942544/6300978111';
  const productionId = 'ca-app-pub-3756584357781172/6027051832';
  // Is a real device and running in production.
  const adUnitID = Constants.isDevice && !__DEV__ ? productionId : testID;

  useEffect(() => {
    if (password === '') {
      return
    }
    const hashedPassword: string = sha1(password)
    const hashRange: string = hashedPassword.slice(0, 5)
    const restOfHash: string = hashedPassword.toUpperCase().slice(5)
    const hashRangeRegEx: RegExp = new RegExp(`${restOfHash}:.*`, 'g')
    fetch(`https://api.pwnedpasswords.com/range/${hashRange}`)
      .then(response => response.text()) // eslint-disable-line
      .then(data => {
        const hit = data.match(hashRangeRegEx)
        if (hit === null) {
          setResults(
            // @ts-ignore
            <Text style={{ color: 'green' }}>
            Not in any known data breaches
            </Text>
          )
        } else {
          const count = hit[0].split(':')[1]
          setResults(
            // @ts-ignore
            <Text style={{ color: 'red' }}>
            This password has been found in {count} breaches!
            </Text>
          )
        }
      })
      .catch(error => console.error(error))
  }, [password])

  return (
    <View style={{ flex: 1 }}>
      <View style={styles.getStartedContainer}>
        <Text
          style={styles.getStartedText}
          lightColor='rgba(0,0,0,0.8)'
          darkColor='rgba(255,255,255,0.8)'
        >
          Has your password been compromised? Find out!
        </Text>
      </View>

      <View style={styles.helpContainer}>
        <TouchableOpacity>
          <TextInput
            style={styles.helpLinkText}
            lightColor={Colors.light.tint}
            onChangeText={(value: string) => {
              setPassword(value)
            }}
            autoCapitalize='none'
            autoCorrect={false}
            placeholder='Password'
            placeholderTextColor='gray'
          />
        </TouchableOpacity>
      </View>

      <View style={{ flex: 1, alignContent: 'center', justifyContent: 'center' }}>
        <Text
          style={styles.resultsText}
          lightColor='rgba(0,0,0,0.8)'
          darkColor='rgba(255,255,255,0.8)'
        >
          {results}
        </Text>
      </View>
      <View>
        <Text style={{ textAlign: 'center', marginBottom: 75 }}>
          Breach Checker will never store or transmit passwords.
          A process called k-anonymity is used to keep your data safe.
        </Text>
        <AdMobBanner
          bannerSize="smartBannerPortrait"
          adUnitID={adUnitID}
          servePersonalizedAds />
      </View>
    </View>
  )
}

const styles = StyleSheet.create({
  getStartedContainer: {
    alignItems: 'center',
    marginHorizontal: 50
  },
  resultsText: {
    fontSize: 17,
    lineHeight: 24,
    textAlign: 'center',
    marginTop: 10,
  },
  getStartedText: {
    fontSize: 17,
    lineHeight: 24,
    textAlign: 'center'
  },
  helpContainer: {
    marginTop: 15,
    marginBottom: 20,
    marginHorizontal: 20,
    alignItems: 'center'
  },
  helpLinkText: {
    textAlign: 'center',
    width: 250,
    height: 60,
    borderColor: 'gray',
    borderBottomWidth: 1.5
  }
})
