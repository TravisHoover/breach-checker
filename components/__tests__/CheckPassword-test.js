import * as React from 'react'
import renderer from 'react-test-renderer'

import CheckPassword from '../CheckPassword';

it('renders correctly', () => { // eslint-disable-line
  const tree = renderer.create(<CheckPassword />).toJSON()

  expect(tree).toMatchSnapshot() // eslint-disable-line
})
