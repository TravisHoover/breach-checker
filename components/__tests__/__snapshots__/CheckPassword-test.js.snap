// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#ffffff",
      },
      Object {
        "flex": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ffffff",
        },
        Object {
          "alignItems": "center",
          "marginHorizontal": 50,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0,0,0,0.8)",
          },
          Object {
            "fontSize": 17,
            "lineHeight": 24,
            "textAlign": "center",
          },
        ]
      }
    >
      Has your password been compromised? Find out!
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ffffff",
        },
        Object {
          "alignItems": "center",
          "marginBottom": 20,
          "marginHorizontal": 20,
          "marginTop": 15,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        onChangeText={[Function]}
        placeholder="Password"
        placeholderTextColor="gray"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "color": "#2f95dc",
            },
            Object {
              "borderBottomWidth": 1.5,
              "borderColor": "gray",
              "height": 60,
              "textAlign": "center",
              "width": 250,
            },
          ]
        }
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ffffff",
        },
        Object {
          "alignContent": "center",
          "flex": 1,
          "justifyContent": "center",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0,0,0,0.8)",
          },
          Object {
            "fontSize": 17,
            "lineHeight": 24,
            "marginTop": 10,
            "textAlign": "center",
          },
        ]
      }
    >
      
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ffffff",
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
          },
          Object {
            "marginBottom": 75,
            "textAlign": "center",
          },
        ]
      }
    >
      Breach Checker will never store or transmit passwords. A process called k-anonymity is used to keep your data safe.
    </Text>
    <View>
      <ViewManagerAdapter_ExpoAdsAdMobBannerView
        proxiedProperties={
          Object {
            "adUnitID": "ca-app-pub-3940256099942544/6300978111",
            "additionalRequestParams": Object {},
            "bannerSize": "smartBannerPortrait",
            "onAdViewDidDismissScreen": undefined,
            "onAdViewDidReceiveAd": undefined,
            "onAdViewWillDismissScreen": undefined,
            "onAdViewWillLeaveApplication": undefined,
            "onAdViewWillPresentScreen": undefined,
            "onDidFailToReceiveAdWithError": [Function],
            "onSizeChange": [Function],
          }
        }
        style={Object {}}
      />
    </View>
  </View>
</View>
`;
